### ISSUE TITLE
JupyterHub 0.8.1 on CentOS 7.4: SystemdSpawner failing to start (500 error, [status=3])

### ISSUE THREAD

**Issue by renglandatsmu** [jupyterhub/jupyterhub#1914]
I am trying to configure a JupyterHub (0.8.1) for a class that will be starting in less than two weeks. The configuration is almost identical (except for the hostname and IP address) to one I configured about five months ago, which continues to work without issue. What is different is that this new server is running CentOS 7.4, not 7.3, with systemd version 219-57 rather than 219-42. I am using SystemdSpawner 0.9.10 on both.

The Hub appears to initialize without issue:

    [I 2018-05-23 15:36:20.635 JupyterHub app:1544] Starting managed service cull-idle
    [I 2018-05-23 15:36:20.636 JupyterHub service:266] Starting service 'cull-idle': ['python', '/apps/jupyterhub/extra/cull_idle_servers.py', '--timeout=86400']
    [I 2018-05-23 15:36:20.643 JupyterHub service:109] Spawning python /apps/jupyterhub/extra/cull_idle_servers.py --timeout=86400
    [D 2018-05-23 15:36:20.647 JupyterHub spawner:727] Polling subprocess every 30s
    [D 2018-05-23 15:36:20.650 JupyterHub proxy:552] Proxy: Fetching GET http://192.168.36.212:5432/api/routes
    [W 2018-05-23 15:36:20.658 JupyterHub proxy:304] Adding missing default route
    [I 2018-05-23 15:36:20.658 JupyterHub proxy:370] Adding default route for Hub: / => http://192.168.36.212:54321
    [D 2018-05-23 15:36:20.658 JupyterHub proxy:552] Proxy: Fetching POST http://192.168.36.212:5432/api/routes/
    [I 2018-05-23 15:36:20.666 JupyterHub app:1581] JupyterHub is now running at https://192.168.36.212:8443/
    [I 2018-05-23 15:36:20.825 JupyterHub log:122] 200 GET /hub/api/users (cull-idle@192.168.36.212) 8.25ms

Here is what JupyterHub logs immediately after I authenticate as user 'joeblow':

    [I 2018-05-23 15:38:17.302 JupyterHub log:122] 302 GET / → /hub (@192.168.36.212) 0.95ms
    [W 2018-05-23 15:38:17.323 JupyterHub base:209] Invalid cookie token
    [W 2018-05-23 15:38:17.326 JupyterHub base:209] Invalid cookie token
    [I 2018-05-23 15:38:17.326 JupyterHub log:122] 302 GET /hub → /hub/login (@192.168.36.212) 4.49ms
    [D 2018-05-23 15:38:17.342 JupyterHub base:268] Setting cookie for joeblow: jupyter-hub-token, {'httponly': True, 'secure': True}
    [I 2018-05-23 15:38:17.344 JupyterHub log:122] 302 GET /hub/login → /hub/home (@192.168.36.212) 5.71ms
    [I 2018-05-23 15:38:17.407 JupyterHub log:122] 200 GET /hub/home (joeblow@192.168.36.212) 50.70ms
    [D 2018-05-23 15:38:17.449 JupyterHub log:122] 200 GET /hub/logo (@192.168.36.212) 8.39ms

But when I click 'Start My Server', the following is logged, and I am presented with the error:
**500 : Internal Server Error**
**Spawner failed to start [status=3]. The logs for joeblow may contain details.**
**You can try restarting your server from the home page.**

    [I 2018-05-23 15:38:38.634 JupyterHub log:122] 302 GET /hub/spawn → /user/joeblow/ (joeblow@192.168.36.212) 4.73ms
    [I 2018-05-23 15:38:38.647 JupyterHub log:122] 302 GET /user/joeblow/ → /hub/user/joeblow/ (@192.168.36.212) 0.80ms
    [D 2018-05-23 15:38:38.663 JupyterHub base:421] Initiating spawn for joeblow
    [D 2018-05-23 15:38:38.702 JupyterHub systemdspawner:132] Bad message (ValueError("unsupported format character 'U' (0x55) at index 7",)): {'name': 'JupyterHub', 'msg': 'user:% Using port %s to start spawning for user %s', 'args': ('joeblow', 52453), 'levelname': 'DEBUG', 'levelno': 10, 'pathname': '/apps/anaconda/3/current/lib/python3.6/site-packages/systemdspawner/systemdspawner.py', 'filename': 'systemdspawner.py', 'module': 'systemdspawner', 'exc_info': None, 'exc_text': None, 'stack_info': None, 'lineno': 132, 'funcName': 'start', 'created': 1527107918.7026882, 'msecs': 702.6882171630859, 'relativeCreated': 139381.61849975586, 'thread': 140040732886848, 'threadName': 'MainThread', 'processName': 'MainProcess', 'process': 16123, 'message': 'Bad message (ValueError("unsupported format character \'U\' (0x55) at index 7",)): {\'name\': \'JupyterHub\', \'msg\': \'user:% Using port %s to start spawning for user %s\', \'args\': (\'joeblow\', 52453), \'levelname\': \'DEBUG\', \'levelno\': 10, \'pathname\': \'/apps/anaconda/3/current/lib/python3.6/site-packages/systemdspawner/systemdspawner.py\', \'filename\': \'systemdspawner.py\', \'module\': \'systemdspawner\', \'exc_info\': None, \'exc_text\': None, \'stack_info\': None, \'lineno\': 132, \'funcName\': \'start\', \'created\': 1527107918.7026882, \'msecs\': 702.6882171630859, \'relativeCreated\': 139381.61849975586, \'thread\': 140040732886848, \'threadName\': \'MainThread\', \'processName\': \'MainProcess\', \'process\': 16123}', 'asctime': '2018-05-23 15:38:38', 'color': '', 'end_color': ''}
    [I 2018-05-23 15:38:38.745 JupyterHub systemdspawner:148] user:joeblow Unit jupyter-joeblow-singleuser in failed state, resetting
    [D 2018-05-23 15:38:38.753 JupyterHub systemdspawner:228] user:joeblow Running systemd-run with: /usr/bin/systemd-run --unit jupyter-joeblow-singleuser --uid 507004 --gid 507004 --setenv=PATH=/apps/anaconda/3/current/bin:/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/root/bin --setenv=LANG=en_US.UTF-8 --setenv=JUPYTERHUB_API_TOKEN=bb05074530824815a2d78f56d7155841 --setenv=JPY_API_TOKEN=bb05074530824815a2d78f56d7155841 --setenv=JUPYTERHUB_CLIENT_ID=user-joeblow --setenv=JUPYTERHUB_HOST= --setenv=JUPYTERHUB_OAUTH_CALLBACK_URL=/user/joeblow/oauth_callback --setenv=JUPYTERHUB_USER=joeblow --setenv=JUPYTERHUB_API_URL=http://192.168.36.212:54321/hub/api --setenv=JUPYTERHUB_BASE_URL=/ --setenv=JUPYTERHUB_SERVICE_PREFIX=/user/joeblow/ --setenv=MEM_LIMIT=4294967296 --setenv=SHELL=/bin/bash --property=MemoryAccounting=yes --property=MemoryLimit=4294967296 /bin/bash -c cd /home/joeblow && exec jupyterhub-singleuser --port=52453 '--notebook-dir="/"' '--NotebookApp.default_url="/tree/home/joeblow"' --debug --disable-user-config
    [D 2018-05-23 15:38:38.782 JupyterHub systemdspawner:253] user:joeblow unit jupyter-joeblow-singleuser is active
    [D 2018-05-23 15:38:38.788 JupyterHub spawner:727] Polling subprocess every 30s
    [D 2018-05-23 15:38:38.799 JupyterHub base:427] 0/100 concurrent spawns
    [D 2018-05-23 15:38:38.800 JupyterHub base:430] 0 active servers
    [D 2018-05-23 15:38:48.816 JupyterHub systemdspawner:256] user:joeblow unit jupyter-joeblow-singleuser is not active
    [W 2018-05-23 15:38:48.819 JupyterHub web:1588] 500 GET /hub/user/joeblow/ (192.168.36.212): Spawner failed to start [status=3]. The logs for joeblow may contain details.
    [D 2018-05-23 15:38:48.824 JupyterHub base:633] No template for 500
    [E 2018-05-23 15:38:48.839 JupyterHub log:114] {
          "X-Forwarded-Proto": "https",
          "X-Forwarded-Port": "443",
          "Connection": "close",
          "X-Forwarded-Server": "jupyter.physics.smu.edu",
          "X-Forwarded-Host": "jupyter.physics.smu.edu",
          "X-Forwarded-For": "10.10.0.88,192.168.36.212",
          "Remote_user": "joeblow",
          "Upgrade-Insecure-Requests": "1",
          "Cookie": "jupyter-hub-token=\"2|1:0|10:1527107897|17:jupyter-hub-token|44:ZGFmNDBhYjRiMWY4NDlhNDkxZWU5OTI4YWZkNTcyYmM=|a9a2a1a213e8a9eb64c38f99fdbca5931a1e7301ed1292d5d664f9dac01eb090\"; _ga=GA1.2.890091753.1497643783; _ga=GA1.3.890091753.1497643783; __utma=212208075.890091753.1497643783.1522687273.1526492922.2; __utmz=212208075.1522687273.1.1.utmcsr=google|utmccn=(organic)|utmcmd=organic|utmctr=(not%20provided); _xsrf=2|3c593266|f68771d8a0bd9c56bb1aafebf40cf4a7|1526663031; _gid=GA1.2.1917826062.1526911395; _shibsession_64656661756c7468747470733a2f2f6a7570797465722e706879736963732e736d752e6564752f73686962626f6c657468=_df25c31227589689ea87093dfa5cf6ca",
          "Referer": "https://jupyter.physics.smu.edu/hub/home",
          "Accept-Encoding": "gzip, deflate, br",
          "Accept-Language": "en-US,en;q=0.5",
          "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
          "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:59.0) Gecko/20100101 Firefox/59.0",
          "Host": "jupyter.physics.smu.edu"
        }
    [E 2018-05-23 15:38:48.840 JupyterHub log:122] 500 GET /hub/user/joeblow/ (joeblow@192.168.36.212) 10177.24ms

And then, after 30 seconds, it logs this:

    [W 2018-05-23 15:39:08.474 JupyterHub user:458] joeblow's server never showed up at http://127.0.0.1:52453/user/joeblow/ after 30 seconds. Giving up
    [D 2018-05-23 15:39:08.483 JupyterHub systemdspawner:256] user:joeblow unit jupyter-joeblow-singleuser is not active
    [E 2018-05-23 15:39:08.497 JupyterHub gen:914] Exception in Future <tornado.concurrent.Future object at 0x7f5db5de0d68> after timeout
        Traceback (most recent call last):
          File "/apps/anaconda/3/current/lib/python3.6/site-packages/tornado/gen.py", line 910, in error_callback
            future.result()
          File "/apps/anaconda/3/current/lib/python3.6/site-packages/jupyterhub/handlers/base.py", line 445, in finish_user_spawn
            yield spawn_future
          File "/apps/anaconda/3/current/lib/python3.6/site-packages/jupyterhub/user.py", line 476, in spawn
            raise e
          File "/apps/anaconda/3/current/lib/python3.6/site-packages/jupyterhub/user.py", line 450, in spawn
            resp = yield server.wait_up(http=True, timeout=spawner.http_timeout)
          File "/apps/anaconda/3/current/lib/python3.6/site-packages/jupyterhub/utils.py", line 180, in wait_for_http_server
            timeout=timeout
          File "/apps/anaconda/3/current/lib/python3.6/site-packages/jupyterhub/utils.py", line 135, in exponential_backoff
            raise TimeoutError(fail_message)
        TimeoutError: Server at http://127.0.0.1:52453/user/joeblow/ didn't respond in 30 seconds

Does anyone know what is happening here? What does the status=3 mean in the Internal Server Error when SystemdSpawner fails?

Thank you.

**Comment by betatim** [comment]
On line `2018-05-23 15:38:38.702` of the logs you see an error message about string formatting from the systemdspawner. In the `msg` there is a spurious % symbol (`user:% Using port %s to start spawning for us`). Not sure if this causes the spawn to fail but it is the place I would start searching.

**Comment by minrk** [comment]
Can you get the logs for joeblow's server? I think it should be something like:

    journalctl -u jupyter-joeblow-singleuser

My guess is that's where the useful info will be.

**Comment by renglandatsmu** [comment]
No, I wasn't able to reference the logs for the singleuser servers at the time I initially posted. However, since that time I have been able to get my Hub working for me. I'm still not completely certain what was causing the issue, but I ended up updating every package (rpm, pip, and conda) to the latest, greatest versions available. Once I'd done that, I was able to start a singleuser server. I was unable to view the contents of my home directory, but I found a couple of other discussions in which it was stated that there is a bug that breaks the c.SystemdSpawner.user_workingdir if it is set, so once I commented that setting out and relocated the home directories (these were formerly in /users, not /home), everything seemed to be working without issue.

Thanks, betatim and minrk, for your efforts to assist me.

Cheers!
